### PR TITLE
vhost-user: Support vhost-user socket reconnection

### PR DESCRIPTION
**Pull request by zhangjaycee**
Hi, all

We implement vhost-user reconnection feature in this patch. The
reconnection here means cloud-hypervisor can reconnect to a new
vhost-user daemon after the old one crashes. 

This patch also support starting a vhost-user daemon after the
start cloud-hypervisor process.

We are also planning to implement vhost-user inflight I/O tracking
based on this feature. Looking forward to any comments. Thanks!

The commit message of this patch is following:

We epoll the EPOLLHUP events for each vhost-user devices in their
epoll thread. When a socket is broken, we notify the vmm thread by
the activate_evt eventfd. Subsequent reconnection will be done in the
activate() procedure of the disconnected vhost-user device.

Vhost-user-blk and vhost-user-fs is already done, while vhost-user-net
is not support yet. Because until this commit inflight virtqueue I/O
tracking is not implemented, this reconnection feature only guarantees
no error if there is no inflight I/O when the socket was broken for
vhost-user-blk. For virtio-fs, remount should be done after
reconnection.

Signed-off-by: Jiachen Zhang <zhangjiachen.jaycee@bytedance.com>

### CODE REVIEW

File: virtio-devices/src/epoll_helper.rs [hunk 1]
```diff
@@ -102,8 +127,11 @@ impl EpollHelper {
             };
 
             for event in events.iter().take(num_events) {
+                let ev_events = event.events as i32;
                 let ev_type = event.data as u16;
 
+                error!("event detected! type: {} event:{}", ev_type, ev_events);
```
Comment (liuw) [inline virtio-devices/src/epoll_helper.rs:133]: This looks to be a debug aid. It should be either dropped or turned into `debug!`

File: virtio-devices/src/vhost_user/blk.rs [hunk 2]
```diff
@@ -37,18 +39,32 @@ impl VhostUserMasterReqHandler for SlaveReqHandler {}
...
-            .map_err(Error::VhostUserCreateMaster)?;
+    pub fn new(
+        id: String,
+        vu_cfg: VhostUserConfig,
+        seccomp_action: SeccompAction,
+        activate_evt: EventFd,
+    ) -> Result<Blk> {
+        let mut vhost_user_blk = retry_with_index(delay::Fixed::from_millis(500), |retry_times| {
+            if retry_times == 1 {
+                println!("waiting for vhost-user-blk deamon opening...");
```
Comment (liuw) [inline virtio-devices/src/vhost_user/blk.rs:63]: Typo "deamon".

Please change `println!` to either `debug!` or `info!`.

"Waiting for vhost-user-blk daemon to open..."

File: virtio-devices/src/vhost_user/blk.rs [hunk 3]
```diff
@@ -261,10 +380,20 @@ impl VirtioDevice for Blk {
                 pause_evt,
                 vu_interrupt_list: interrupt_list_sub,
                 slave_req_handler: None,
+                sockfd: Some(self.vhost_user_blk.as_mut().unwrap().as_raw_fd()),
```
Comment (liuw) [inline virtio-devices/src/vhost_user/blk.rs:383]: Is `as_mut` here really needed? Can you replace it with `as_ref`? This comment applies to other places where `as_raw_fd` is called.

File: virtio-devices/src/device.rs [hunk 4]
```diff
@@ -197,6 +197,13 @@ pub trait VirtioDevice: Send {
             offset_config.write_all(data).unwrap();
         }
     }
+
+    /// Default implementation if the device is not vhost-user or reconnection is
+    /// not supported, disconnected status will always be false for this default
+    /// one.
+    fn get_disconnected_arc(&self) -> Result<Arc<AtomicBool>, std::io::Error> {
+        Ok(Arc::new(AtomicBool::new(false)).clone())
```
Comment (liuw) [inline virtio-devices/src/device.rs:205]: There is no need to call `clone`, is there?

File: virtio-devices/src/vhost_user/blk.rs [hunk 5]
```diff
@@ -261,10 +380,20 @@ impl VirtioDevice for Blk {
...
+                disconnected: Some(self.disconnected.clone()),
+                activate_evt: Some(
+                    self.activate_evt
+                        .try_clone()
+                        .map_err(ActivateError::CloneEventfd)
```
Comment (liuw) [inline virtio-devices/src/vhost_user/blk.rs:388]: Don't you want to use `?` to return an error?

File: virtio-devices/src/vhost_user/blk.rs [hunk 6]
```diff
@@ -215,21 +325,30 @@ impl VirtioDevice for Blk {
...
-            queues,
+            queues_real,
             queue_evts,
             &interrupt_cb,
             self.common.acked_features,
         )
         .map_err(ActivateError::VhostUserBlkSetup)?;
 
-        let mut epoll_threads = Vec::new();
+        let epoll_threads = Vec::new();
```
Comment (liuw) [inline virtio-devices/src/vhost_user/blk.rs:351]: Why is this changed?

File: virtio-devices/src/vhost_user/blk.rs [hunk 7]
```diff
@@ -261,10 +380,20 @@ impl VirtioDevice for Blk {
...
+                    self.activate_evt
+                        .try_clone()
+                        .map_err(ActivateError::CloneEventfd)
+                        .unwrap(),
+                ),
             });
 
             let paused = self.common.paused.clone();
             let paused_sync = self.common.paused_sync.clone();
+            let mut epoll_threads = Vec::new();
```
Comment (liuw) [inline virtio-devices/src/vhost_user/blk.rs:395]: Why is this here?

You essentially moved the previous `let mut epoll_threads = ...` here.

File: virtio-devices/src/vhost_user/fs.rs [hunk 8]
```diff
@@ -482,10 +501,24 @@ impl VirtioDevice for Fs {
...
+                self.activate_evt
+                    .try_clone()
+                    .map_err(ActivateError::CloneEventfd)
+                    .unwrap(),
+            ),
         });
 
         let paused = self.common.paused.clone();
         let paused_sync = self.common.paused_sync.clone();
+        /*
```
Comment (liuw) [inline virtio-devices/src/vhost_user/fs.rs:516]: Please drop this comment block. Unused code should be deleted before submitting.

File: virtio-devices/src/vhost_user/blk.rs [hunk 9]
```diff
@@ -37,18 +39,32 @@ impl VhostUserMasterReqHandler for SlaveReqHandler {}
...
+        seccomp_action: SeccompAction,
+        activate_evt: EventFd,
+    ) -> Result<Blk> {
+        let mut vhost_user_blk = retry_with_index(delay::Fixed::from_millis(500), |retry_times| {
+            if retry_times == 1 {
+                println!("waiting for vhost-user-blk deamon opening...");
+            }
+            Master::connect(&vu_cfg.socket, vu_cfg.num_queues as u64)
+        })
+        .unwrap();
```
Comment (liuw) [inline virtio-devices/src/vhost_user/blk.rs:67]: Is there a way to return a sensible error? Or are we better off panicking the whole program if vhost-user-blk is not reachable? What is your thought here?

**Review by liuw — COMMENTED** [review]
I skimmed through the code and made a bunch of comments. I see you added some code for handling HUP but couldn't figure out how it interacts with the daemon or the main process. Unfortunately my time ran out before I could dig deeper.

I think this PR can benefit from the following things:

1. A clear description on what it tries to achieve and how it achieves its goal.
2. Break it up into several patches. One for refactoring, one for introducing (new) core mechanism, the one for each type of backend that makes use of the mechanism.

**Comment by zhangjaycee** [comment]
> I skimmed through the code and made a bunch of comments. I see you added some code for handling HUP but couldn't figure out how it interacts with the daemon or the main process. Unfortunately my time ran out before I could dig deeper.
> 
> I think this PR can benefit from the following things:
> 
> 1. A clear description on what it tries to achieve and how it achieves its goal.
> 2. Break it up into several patches. One for refactoring, one for introducing (new) core mechanism, the one for each type of backend that makes use of the mechanism.

Thanks for the review and helpful comments!  We will split this patch and rework the commit messages.

I'd like to first briefly describe the more detailed mechanism here. When a HUP event is polled by a vhost-user device's epoll thread, the epoll thread will notify the vmm thread by the `activate_evt` eventfd. The vmm thread then will call the`activate()` function of the disconnected vhost-user device. And as we move the socket `Master::connect()` from `Fs::new()` and `Blk::new()` to `Fs::activate()` and `Blk::new()`, the new socket connect can be established again. The `Master::connect()` is also wrapped by a retry loop, which is used to wait for the new daemon to start.

File: virtio-devices/src/epoll_helper.rs (hunk 1 above)
Comment (zhangjaycee) [reply virtio-devices/src/epoll_helper.rs]: I removed this line.

File: virtio-devices/src/vhost_user/blk.rs (hunk 2 above)
Comment (zhangjaycee) [reply virtio-devices/src/vhost_user/blk.rs]: I changed `println!` to `info!`.

File: virtio-devices/src/vhost_user/blk.rs (hunk 3 above)
Comment (zhangjaycee) [reply virtio-devices/src/vhost_user/blk.rs]: I changed them to `as_ref`.

File: virtio-devices/src/device.rs (hunk 4 above)
Comment (zhangjaycee) [reply virtio-devices/src/device.rs]: I removed the `.clone()`.

File: virtio-devices/src/vhost_user/blk.rs (hunk 5 above)
Comment (zhangjaycee) [reply virtio-devices/src/vhost_user/blk.rs]: I think you are right, I changed `.unwrap()` to `?`.

File: virtio-devices/src/vhost_user/blk.rs (hunk 7 above)
Comment (zhangjaycee) [reply virtio-devices/src/vhost_user/blk.rs]: I removed this line and keep the `mut` mentioned in your previous comment.

File: virtio-devices/src/vhost_user/fs.rs (hunk 8 above)
Comment (zhangjaycee) [reply virtio-devices/src/vhost_user/fs.rs]: I removed it.

File: virtio-devices/src/vhost_user/blk.rs (hunk 9 above)
Comment (zhangjaycee) [reply virtio-devices/src/vhost_user/blk.rs]: I replaced `unwrap()` by a `map_err()?`.

**Review by sboeuf — CHANGES_REQUESTED** [review]
I'm not in favor of this design. The fact that we reuse the activate code path is wrong because the activate code path is only meant to be triggered when the driver is ready, or when the device is reset.
The approach I'd like to take here would be to keep everything internal to the vhost-user device (fs, blk, net). That means upon detection of `HUP`, we would reconnect and go through the vhost-user setup again. This will require the `Fs`, `Blk`, `Net` to share some structures with `VhostUserEpollHandler`.

If we take `virtio-devices/src/vhost_user/fs.rs` as example, here's what we could do:
- Define a trait `VhostUserReconnect` that would provide one method called `reconnect()` or `initialize()`.
- Declare a structure `FsReconnect` that would contain all the information needed to perform a proper reconnection/setup from scratch.
- Implement `VhostUserReconnect` for `FsReconnect`
- Provide the trait `VhostUserReconnect` to `VhostUserEpollHandler`
- Upon HUP event, `VhostUserEpollHandler` would use `VhostUserReconnect` trait to call into `reconnect()`

**Comment by sboeuf** [comment]
No progress on this PR, let's close it. Feel free to open a new one when ready.

**Comment by sboeuf** [comment]
@zhangjaycee any update on this?
I'm trying to understand why you need this feature for? Do you have a use case in mind (OVS-DPDK maybe?).
I've been doing some research on the reconnection for the client mode, and I would like to understand if that is something that might be appropriate for your use case?

**Comment by zhangjaycee** [comment]
> @zhangjaycee any update on this?
> I'm trying to understand why you need this feature? Do you have a use case in mind (OVS-DPDK maybe?).
> I've been doing some research on the reconnection for the client mode, and I would like to understand if that is something that might be appropriate for your use case?

Thanks, Sebastien! I am sorry there is no update yet. But we are planning to revise the patch the next month.

The main use case now is the crash recovery of virtiofsd and vhost-user-blk daemon. Crash reconnection of OVS-DPDK might also rely on this feature but currently is not our use case. Besides, it is also very possible to support the live-upgrade feature of the vhost-user daemons based on the crash recovery feature.  I also replied to your mail (https://www.mail-archive.com/qemu-devel@nongnu.org/msg806270.html) in the virtio-fs mail list, I listed some background context related to this PR in the mail.

It should be noted that to support vhost-user daemon crash recovery, the socket reconnection feature in the PR alone can not completely solve the problem. From the vhost-user perspective, inflight virtqueue I/O requests should also be tracked and restored. From the vhost-user daemons perspective, virtiofsd, vhost-user-blk daemon, or OVS-DPDK also have their own internal status. So in the future, we also need to expand the vhost-user implementation of cloud-hypervisor to support virtqueue inflight I/O tracking protocols (VHOST_USER_PROTOCOL_F_INFLIGHT_SHMFD). The status of vhost-user daemons should be handled respectively in their own implementations.

Jiachen

**Comment by sboeuf** [comment]
@zhangjaycee thanks for clarifying the situation :)
I agree the reconnection will not solve the problem entirely, but we should start with this first. We should make sure we can have the VMM reconnecting to vhost-user backends in case the connection got dropped, or to relisten in case the VMM was acting as the server.
The support for INFLIGHT I/O should happen as a follow up PR, and might require some extra changes to the `rust-vmm/vhost` crate as well.

Before you go ahead and start implementing the solution, could you detail the design you have in mind so that we can agree on it before you start coding?

Last thing, `vhost-user-blk` is simpler than `virtio-fs`, so I would suggest we try to make this one work first, or even `vhost-user-net` since there's not much state that needs to be saved.

**Comment by zhangjaycee** [comment]
> @zhangjaycee thanks for clarifying the situation :)
> I agree the reconnection will not solve the problem entirely, but we should start with this first. We should make sure we can have the VMM reconnecting to vhost-user backends in case the connection got dropped, or to relisten in case the VMM was acting as the server.
> The support for INFLIGHT I/O should happen as a follow up PR, and might require some extra changes to the `rust-vmm/vhost` crate as well.
> 
> Before you go ahead and start implementing the solution, could you detail the design you have in mind so that we can agree on it before you start coding?
>

Good, next week, we will propose a plan for the revision of this patch and the following up inflight patches.

> 
> Last thing, `vhost-user-blk` is simpler than `virtio-fs`, so I would suggest we try to make this one work first, or even `vhost-user-net` since there's not much state that needs to be saved.

That's fine, we can start with `vhost-user-blk`. And we are also trying to handle the internal status of virtiofsd, so that virtiofsd can be recovered without extra interactions with QEMU or cloud-hypervisor other than the "socket reconnection" and the "inflight I/O protocol". 

Thank!
Jiachen

**Comment by zhangjaycee** [comment]
> I'm not in favor of this design. The fact that we reuse the activate code path is wrong because the activate code path is only meant to be triggered when the driver is ready, or when the device is reset.
> The approach I'd like to take here would be to keep everything internal to the vhost-user device (fs, blk, net). That means upon detection of `HUP`, we would reconnect and go through the vhost-user setup again. This will require the `Fs`, `Blk`, `Net` to share some structures with `VhostUserEpollHandler`.
> 
> If we take `virtio-devices/src/vhost_user/fs.rs` as example, here's what we could do:
> 
> * Define a trait `VhostUserReconnect` that would provide one method called `reconnect()` or `initialize()`.
> * Declare a structure `FsReconnect` that would contain all the information needed to perform a proper reconnection/setup from scratch.
> * Implement `VhostUserReconnect` for `FsReconnect`

Hi, Sebastien,

Thanks for the suggestion. I reconsidered the code execution paths again, I think your idea is reasonable and elegant. The problem now is, actually,  we should do most of the `activate()` work when reconnecting. In this patch, take the virtio-fs device as an example, the socket (re)connection code is moved from `fs::new()` to `fs::activate()`. If we declare a new `struct FsReconnect`, I'm afraid finally, we should move almost all `struct Fs` fields into it. So maybe we could rename the `struct FsReconnect` you mentioned to `struct FsInternal`. And this struct should be wrapped by `Arc` as it should be shared with the `VhostUserEpollHandler` thread for socket (re)connection.

> * Provide the trait `VhostUserReconnect` to `VhostUserEpollHandler`
> * Upon HUP event, `VhostUserEpollHandler` would use `VhostUserReconnect` trait to call into `reconnect()`

In cloud-hypervisor master, the `VhostUserEpollHandler` thread is created in `fs::activate()`. And in this patch, when an `EPOLLHUP` event is detected, the thread will set the `disconnected` flag, then tell the main `Vmm` thread to `activate_virtio_devices()` by writing the `activate_evt` eventfd, and finally terminates. A new `VhostUserEpollHandler` thread will be created by the `fs::activate()` and to watch the events for the new connection. If the `VhostUserEpollHandler` takes the responsibility to (re)connect (rather than use the vmm main thread in this patch), I think the `VhostUserEpollHandler` thread may need to outlive the `Fs::FsInternal`. 

Therefore, maybe we should spawn the `VhostUserEpollHandler` thread in `Fs::new()`,  pass the `Fs::FsInternal` field into the thread, and both `activate()` and the new `reconnect()` function should be handled in that thread. The `activate()` function and `reconnect()` function would share some code path, which could be the `initialize()`. Could this be our next step of this feature? 

Further, as the `activate()` will be handled in the respective `VhostUserEpollHandler` thread of each vhost-user device, the `activate_evt` eventfd can also separately handled in each thread of the vhost-user devices, rather than in the main `vmm` thread.

As for the inflight I/O tracking, other than the `vhost` crate, we also need to change the code of `vm-virtio`. I notice `vhost` is maintained in `rust-vmm/vhost`, while `vm-virtio` is maintained in this repository. I think both of them need some code changes.

All the best,
Jiachen

**Comment by sboeuf** [comment]
Hi @zhangjaycee , thanks for sharing your thoughts on the topic :)
The `virtio-fs` device is the most complex when compared with `vhost-user-net` and `vhost-user-blk`. I have a simple design in mind that I'd like to experiment first, and if it works, we could try to adapt it to make it work for `virtio-fs` as well.

The idea is to assume that we won't handle disconnection/reconnection until the device is activated, this will simplify things a lot since we can store all the information at once, with no need for `Option<>` fields. I was thinking that after `setup_vhost_user()` gets called, we could spawn a dedicated thread for detecting and handling the reconnection. This new thread would listen on the connection, and in case a `HUP` is detected, it would try to reconnect for a minute. In case the reconnection succeeds, and because we assume the guest shouldn't know about it, we can simply rely on the previously negotiated features to set them (both virtio and protocol features), along with the queues and event file descriptors.

So to summarize, I'm proposing a very simple approach with a new thread taking care of detecting a disconnection, reconnecting to the backend, and then setting things up once again.
In case of virtio-fs, this solution will be slightly more complex as it needs to set the SLAVE_REQ file descriptor as well.

Can you think of any blocker with this approach?

**Comment by sboeuf** [comment]
@zhangjaycee here is the PR I was talking about: https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2694
Please have a look at the design, which works fine for `vhost-user-net`. I hope we can reuse the same idea for `virtio-fs` and `vhost-user-blk`.

**Comment by zhangjaycee** [comment]
> @zhangjaycee here is the PR I was talking about: #2694
> Please have a look at the design, which works fine for `vhost-user-net`. I hope we can reuse the same idea for `virtio-fs` and `vhost-user-blk`.

Great! I'll take a look and maybe try to implement the design for `virtio-fs` and `vhost-user-blk`. 

Jiachen

**Comment by sboeuf** [comment]
@zhangjaycee 
>Great! I reviewed the PR, and it looks good to me.

Thank you, I really appreciate you taking the time to review it!

>I think I can reuse the shared code path in your PR to support vhost-user-blk reconnect first.
I will start to develop the vhost-user-blk part based on your PR commits, and rebase the changes to master after the vhost-user-net part and the shared part is merged. Let me know if you have plans or already start to develop vhost-user-blk or virtio-fs part.

That sounds like a great plan! I haven't started anything on `vhost-user-blk` or `virtio-fs` so I'll let you take care of it if that's okay with you. Please keep me posted on your progress and don't hesitate to share any blocker that you might encounter.
